### PR TITLE
Subsharding vindex document

### PR DESCRIPTION
**Pull request by harshit-gangal**
This adds subsharding vindex documentation with `multicol` vindex.

### CODE REVIEW

File: content/en/docs/13.0/user-guides/vschema-guide/subsharding-vindex.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
+---
+title: Subsharding Vindex
+weight: 11
+---
+
+### Introduction
+
+A subsharding vindex is a functional unique vindex. It is represented by using a set of columns from the table. 
+If all the columns are provided in query routing using where clause then it is routed to single shard.
+If subset of columns in order of the appearance are provided then the qiuery is routed to subset of shards instead of all shards.
```
Comment (GuptaManan100) [inline content/en/docs/13.0/user-guides/vschema-guide/subsharding-vindex.md]: qiuery to query

File: content/en/docs/13.0/user-guides/vschema-guide/subsharding-vindex.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+### Usage
+
+This vindex is registered as `multicol` vindex.
+
+The Vindex takes in 3 inputs
+1. `column_count` - the number of columns that would be provided for using the vindex.
+2. `column_vindex` - hashing function each column will use to provide hash value for that column
+3. `column_bytes` - bytes to be used from each column's hash value after applying hashing function on it to produce keyspace id.
+
+Usage in VSchema:
```
Comment (GuptaManan100) [inline content/en/docs/13.0/user-guides/vschema-guide/subsharding-vindex.md]: Usage examples in VSchema:

**Review by GuptaManan100 — APPROVED** [review]
Just a few nits, other than that everything looks good to me.

File: content/en/docs/13.0/user-guides/vschema-guide/subsharding-vindex.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+weight: 11
+---
+
+### Introduction
+
+A subsharding vindex is a functional unique vindex. It is represented by using a set of columns from the table. 
+If all the columns are provided in query routing using where clause then it is routed to single shard.
+If subset of columns in order of the appearance are provided then the query is routed to subset of shards instead of all shards.
+
+### Use Case
```
Comment (deepthi) [inline content/en/docs/13.0/user-guides/vschema-guide/subsharding-vindex.md:12]: Use cases

File: content/en/docs/13.0/user-guides/vschema-guide/subsharding-vindex.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+	  }
+    }
+}
+```
+```json
+"tables": {
+   "multicol_tbl": {
+	  "column_vindexes": [
+	    {
+                "columns": ["cola","colb","colc"],
```
Comment (deepthi) [inline content/en/docs/13.0/user-guides/vschema-guide/subsharding-vindex.md:47]: Fix indentation.

File: content/en/docs/13.0/user-guides/vschema-guide/subsharding-vindex.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+A common use case is when a given entity’s data fits on more than one shard (e.g a whale tenant) and there is no suitable intermediary grouping construct.
+Storing that entity’s data across all shards will lead to expensive scatter queries.
+
+This is also applicable for geo-sharding use cases where the data for particular region should recide in subset of shards which are co-located to that region.
+
+### Usage
+
+This vindex is registered as `multicol` vindex.
+
+The Vindex takes in 3 inputs
```
Comment (deepthi) [inline content/en/docs/13.0/user-guides/vschema-guide/subsharding-vindex.md:23]: Remove the new line and continue the paragraph with `It takes 3 parameters as input`.

File: content/en/docs/13.0/user-guides/vschema-guide/subsharding-vindex.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+1. `column_count` - the number of columns that would be provided for using the vindex.
+2. `column_vindex` - hashing function each column will use to provide hash value for that column
```
Comment (deepthi) [inline content/en/docs/13.0/user-guides/vschema-guide/subsharding-vindex.md:25]: provide hash value -> compute the hash value

File: content/en/docs/13.0/user-guides/vschema-guide/subsharding-vindex.md [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+3. `column_bytes` - bytes to be used from each column's hash value after applying hashing function on it to produce keyspace id.
```
Comment (deepthi) [inline content/en/docs/13.0/user-guides/vschema-guide/subsharding-vindex.md:26]: hashing function -> its hashing function

File: content/en/docs/13.0/user-guides/vschema-guide/subsharding-vindex.md [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+	  "column_vindexes": [
+	    {
+                "columns": ["cola","colb","colc"],
+		"name": "multicol_vdx"
+	    }
+      ]
+   }
+}
+```
+`column_count` is the mandatory parameter that needs to be provided.
```
Comment (deepthi) [inline content/en/docs/13.0/user-guides/vschema-guide/subsharding-vindex.md:54]: `column_count` is a required parameter.

File: content/en/docs/13.0/user-guides/vschema-guide/subsharding-vindex.md [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+A maximum of 8 columns can be used in this vindex i.e. `column_count <= 8`
+
+`column_vindex` should contain the vindex name in a comma-separated list. It should be less than equal to column_count.
```
Comment (deepthi) [inline content/en/docs/13.0/user-guides/vschema-guide/subsharding-vindex.md:57]: name -> names
less than equal -> less than or equal

File: content/en/docs/13.0/user-guides/vschema-guide/subsharding-vindex.md [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+Default vindex is `hash` vindex, any column for which vindex is not provided, the default vindex will be used.
```
Comment (deepthi) [inline content/en/docs/13.0/user-guides/vschema-guide/subsharding-vindex.md:58]: The default vindex is `hash`. If a vindex is not provided for a column, then `hash` will be used for that column.

File: content/en/docs/13.0/user-guides/vschema-guide/subsharding-vindex.md [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+Vindex in `column_vindex` should implement the below interface otherwise the initialization will fail.
```
Comment (deepthi) [inline content/en/docs/13.0/user-guides/vschema-guide/subsharding-vindex.md:59]: Each Vindex in `column_vindex`...

File: content/en/docs/13.0/user-guides/vschema-guide/subsharding-vindex.md [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+
+```go
+// Hashing defined the interface for the vindexes that export the Hash function to be used by multi-column vindex.
```
Comment (deepthi) [inline content/en/docs/13.0/user-guides/vschema-guide/subsharding-vindex.md:62]: // Hashing defines the interface for vindexes that export the Hash function to be used by a multi-column vindex.

Ideally this should be fixed in the source code too.

File: content/en/docs/13.0/user-guides/vschema-guide/subsharding-vindex.md [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+type Hashing interface {
+	Hash(id sqltypes.Value) ([]byte, error)
+}
+```
+
+Eg for how keyspace id is created
```
Comment (deepthi) [inline content/en/docs/13.0/user-guides/vschema-guide/subsharding-vindex.md:68]: Here's an example of how the keyspace ID is constructed:

File: content/en/docs/13.0/user-guides/vschema-guide/subsharding-vindex.md [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+c3 - 4 bytes
+
+For the column Cn -> Cn.Hash(Cn_values)[0 : n_bytes_allocated_for_column]
+
+keyspace_id:
+|_c1|_c2__c2__c2|_c3__c3__c3__c3|
+|_0_|_1_|_2_|_3_|_4_|_5_|_6_|_7_|
+```
+
+`column_bytes` should contain bytes in a comma-separated list. The total count should be equal to 8 bytes.
```
Comment (deepthi) [inline content/en/docs/13.0/user-guides/vschema-guide/subsharding-vindex.md:82]: `column_bytes` should contain the number of bytes for each column in a comma-separated list. The numbers should add up to 8.

File: content/en/docs/13.0/user-guides/vschema-guide/subsharding-vindex.md [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+If for some columns bytes are not represented then it is calculated by assigning equal bytes to remaining unassigned columns.
```
Comment (deepthi) [inline content/en/docs/13.0/user-guides/vschema-guide/subsharding-vindex.md:83]: If the number of bytes is not provided for all columns, they are inferred by assigning equal numbers to the remaining unassigned columns.

Comment (deepthi) [reply content/en/docs/13.0/user-guides/vschema-guide/subsharding-vindex.md:83]: There must be some rounding involved here. We should document that.
For example: 3 columns, 8 bytes.
Does that result in 2,3,3 or 3,2,3 or 3,3,2?

Another example: 3 columns, first column is 1 byte.
Will the other two be 3,4 or 4,3?

File: content/en/docs/13.0/user-guides/vschema-guide/subsharding-vindex.md [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+Calculated:
+remaining bytes = 8 - 1 - 3 -> 4
+remaining columns = 5 - 2 -> 3
+col 2 -> 2
+col 4 -> 1
+col 5 -> 1
+```
+
+### Hashing Function Implementation
+Below is the list of vindex that has implemented the hashing interface required to be passed in `column_vindex` for using `multicol` vindex.
```
Comment (deepthi) [inline content/en/docs/13.0/user-guides/vschema-guide/subsharding-vindex.md:106]: Currently the website has:
```
The Vitess vindexes that implement the hashing interface requires and can be used in a subsharding vindex:
```
requires -> required for this phrase to make sense.
can also replace `The Vitess vindexes` with just `Vindexes`

**Review by deepthi — COMMENTED** [review]
Nice guide. I looked at the current version on the website before making these suggestions for improvement.

File: content/en/docs/13.0/user-guides/vschema-guide/subsharding-vindex.md (hunk 15 above)
Comment (harshit-gangal) [reply content/en/docs/13.0/user-guides/vschema-guide/subsharding-vindex.md:83]: There is an example that shows that
Eg:

```
Given:
column_count = 5
column_bytes = 1, , 3

col 1 -> 1
col 2 -> not provided
col 3 -> 3
col 4 -> not provided
col 5 -> not provided

Calculated:
remaining bytes = 8 - 1 - 3 -> 4
remaining columns = 5 - 2 -> 3
col 2 -> 2
col 4 -> 1
col 5 -> 1
```

Comment (harshit-gangal) [reply content/en/docs/13.0/user-guides/vschema-guide/subsharding-vindex.md:83]: added more examples

**Pull request merged**